hashintel/hash
# H-1188, H-1209: Split pages into 'document' and 'canvas'. Fix text saving + other bugs

## 🌟 What is the purpose of this PR?

The primary motivation of this PR is to distinguish between 'document' and 'canvas' pages, so we can type them properly – they have different linking information between the page and the block, to store either an index or spatial position.

As part of this I have updated document pages to be ordered by a fractional string index rather than an integer. This meant updating the save process, and so I have also updated it to remove some bugs related to ordering, as well as fixed some other things as listed under 'What does this change?'.

A fair bit of the diff comes from regenerating system types to use `https://hash.ai` in their property keys (fixes H-1171).

## 🔍 What does this change?



- Fixes H-1188, H-1194: distinguish between `Document` and `Canvas`, which inherit from `Page`. Split the `contains` link into `hasIndexedContent` and `hasSpatiallyPositionedContent`. Add canvas icon.
- Fixes H-1209: rewrite save process to fix the following bugs:
  - moving blocks is not working in `main` (fixes H-1208)
  - nor is removing multiple blocks (and sometimes one block)
  - text was being lost, with the end of paragraphs missing on refresh 
  - there was a crash if a user was viewing a page when another user added a block to it. Fixes H-1069 (we need a follow up H-1260 to have changes appear from other users and to stop them messing with each other's work, but it doesn't crash at least)
- Brings #3564 into `main` as well.

In passing:
- Fix a minor styling bug with mentions in comments (Fixes H-1203)
- Add polling to comments so that comments appear for User B when User A adds one if they are viewing the same page
- Better error handling for the 'fetch BP blocks' API route

## Pre-Merge Checklist 🚀

### 🚢 Has this modified a publishable library?




This PR:



### 📜 Does this require a change to the docs?




The changes in this PR:



### 🕸️ Does this require a change to the Turbo Graph?




The changes in this PR:


## ⚠️ Known issues

Outstanding text editing issues:
- Collaborative editing in general. H-1260
- We are not taking advantage of fractional indexing to minimise entity updates at the moment (see comment). H-1259
- There's a flash in text blocks sometimes when editing, at least locally. I believe this is the block being remounted, unsure if by ProseMirror or React (the content disappears briefly)

## 🛡 What tests cover this?



- Some existing for pages, which I have updated.

## ❓ How to test this?



1. Check that creating pages still works, including canvas (via top right create button)
2. Check pinned profile tab for pages
3. Try text editing and refreshing, moving blocks around etc

## 📹 Demo



https://github.com/hashintel/hash/assets/37743469/4c876d34-62c9-42dd-a77a-3ffa458bef49
